File: core/src/ru/myitschool/dcrawler/entities/ai/GoblinAI.java
package ru.myitschool.dcrawler.entities.ai;

import com.badlogic.gdx.utils.Array;
import ru.myitschool.dcrawler.entities.ai.pathfinding.NodePath;
import ru.myitschool.dcrawler.entities.ai.task.MoveTask;
import ru.myitschool.dcrawler.entities.ai.task.SkillTask;
import ru.myitschool.dcrawler.entities.Entity;
import ru.myitschool.dcrawler.entities.skills.ProtectiveStance;
import ru.myitschool.dcrawler.entities.skills.Scratches;
import ru.myitschool.dcrawler.entities.skills.Target;

import java.util.Map;

/**
 * Created by Voyager on 16.08.2017.
 */
public class GoblinAI extends AI {

    public static final int LONG_DISTANCE_STATE = 0;
    public static final int NORMALL_DISTANCE_STATE = 1;
    public static final int SHORT_DISTANCE_STATE = 2;

    Entity controlledEntity;

    Scratches scratchSkill;
    ProtectiveStance stanceSkill;

    public GoblinAI(Entity controlledEntity) {
        super(controlledEntity);
        this.controlledEntity = controlledEntity;
        scratchSkill = new Scratches(controlledEntity);
        stanceSkill = new ProtectiveStance(controlledEntity);
        controlledEntity.addSkill(scratchSkill);
        controlledEntity.addSkill(stanceSkill);
    }

    @Override
    public void aiAnalyze() {
        Map<NodePath, Entity> paths = AIUtils.getAllEntityPaths(controlledEntity.getTileX(), controlledEntity.getTileY(),false, Entity::isCharacter);
        if (paths.size() > 0) {
            Map.Entry<NodePath, Entity> entityPath = (Map.Entry<NodePath, Entity>) paths.entrySet().toArray()[0];
            Entity entity = entityPath.getValue();
            NodePath path = entityPath.getKey();
            path.cutLast();
            int distance = path.getCost();
            int speed = controlledEntity.getSpeed();
            if (distance > 0 && distance <= controlledEntity.getSpeed()) {
                addTask(new MoveTask(path));
                Array<Target> targets = new Array<Target>();
                targets.add(new Target(entity.getTileX(), entity.getTileY()));
                addTask(new SkillTask(scratchSkill, targets));
            } else if (distance > speed) {
                addTask(new MoveTask(path, true, true));
            } else if (distance == 0) {
                addTask(new SkillTask(stanceSkill, new Target(entity.getTileX(), entity.getTileY())));
            }
        }
    }

    @Override
    public void startTurn() {
        super.startTurn();
    }

    @Override
    public void endMove() {
        super.endMove();
    }

    @Override
    public void endSkill() {
        super.endSkill();
    }
}

File: core/src/com/wolg_vlad/dcrawler/entities/ai/GoblinAI.kt
package com.wolg_vlad.dcrawler.entities.ai

import com.badlogic.gdx.utils.Array
import com.wolg_vlad.dcrawler.entities.Entity
import com.wolg_vlad.dcrawler.entities.ai.task.MoveTask
import com.wolg_vlad.dcrawler.entities.ai.task.SkillTask
import com.wolg_vlad.dcrawler.entities.skills.ProtectiveStance
import com.wolg_vlad.dcrawler.entities.skills.Scratches
import com.wolg_vlad.dcrawler.entities.skills.Target

/**
 * Created by Voyager on 16.08.2017.
 */
class GoblinAI(controlledEntity: Entity) : AI(controlledEntity) {
    var scratchSkill: Scratches
    var stanceSkill: ProtectiveStance
    override fun aiAnalyze() {
        val paths = AIUtils.getAllEntityPaths(controlledEntity.tileX, controlledEntity.tileY, false, Entity::isCharacter)
        if (paths!!.size > 0) {
            val entityPath = paths.entries.toTypedArray()[0]
            val entity = entityPath.value
            val path = entityPath.key
            path.cutLast()
            val distance = path.cost
            val speed: Int = controlledEntity.speed
            if (distance > 0 && distance <= controlledEntity.speed) {
                addTask(MoveTask(path))
                val targets = mutableListOf<Target>()
                targets.add(Target(entity.tileX, entity.tileY))
                addTask(SkillTask(scratchSkill, targets))
            } else if (distance > speed) {
                if (path.cost > entity.speed * 2 + 1) {
                    path.cut(entity.speed * 2 + 1)
                }
                addTask(MoveTask(path))
            } else if (distance == 0) {
                addTask(SkillTask(stanceSkill, Target(entity.tileX, entity.tileY)))
            }
        }
    }

    override fun startTurn() {
        super.startTurn()
    }

    override fun endMove() {
        super.endMove()
    }

    override fun endSkill() {
        super.endSkill()
    }

    companion object {
        const val LONG_DISTANCE_STATE = 0
        const val NORMALL_DISTANCE_STATE = 1
        const val SHORT_DISTANCE_STATE = 2
    }

    init {
        this.controlledEntity = controlledEntity
        scratchSkill = Scratches(controlledEntity)
        stanceSkill = ProtectiveStance(controlledEntity)
        controlledEntity.addSkill(scratchSkill)
        controlledEntity.addSkill(stanceSkill)
    }
}